### PR TITLE
Re-organizing collation utils

### LINKED ISSUE #1261: Organize our collation utils
see #1005 for some context. From @ebsmothers  (and @joecummings)@

> I don't love our collate utilities rn. In an ideal world I want two simple abstractions: right_padded_collate and left_padded_collate, and any further padding built up on top of that. I don't think this version should exist at all (especially with its heavy dictionary usage)

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

closes #1261, ref #1249

#### Changelog
This PR:
- Renames our current collation utility which is used in our SFT recipes to `padded_collate_sft`, to indicate that it is task-specific.
- Changes our current left-padding collation utilities to a generic `left_pad_sequence_function`. 
- Introduces a generic `padded_collate`, which can be flexibly configured with a choice of base padding function to suit tasks not covered by `padded_collate_sft`. This is demonstrated through the PPO recipe currently, but the tests for the function also show how it could be used for e.g. classification tasks.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it. 
- [ ] I did not change any public API;
- [x] I have added an example to docs or docstrings;

### CODE REVIEW

File: torchtune/data/_collate.py [hunk 1]
```diff
@@ -12,7 +12,133 @@
...
+        batch_first=batch_first,
+        padding_value=padding_value,
+    ).flip(dims=[1])
+
+
 def padded_collate(
+    batch: List[Dict[str, List[int]]],
+    *,
+    pad_fn: Callable,
+    keys_to_pad: List[str],
```
Comment (SalmanMohammadi) [inline torchtune/data/_collate.py:60]: I'm not 100% about how this should be paramterized and very open to suggestions. The current setup means all `keys_to_pad` will be collated with `pad_fn`, even if they don't need padding, which shouldn't have any effect when the key doesn't need padding e.g. in the case of classification tasks, but it's not super intuitive.

Comment (SalmanMohammadi) [reply torchtune/data/_collate.py:60]: After looking back on this, it looks silly. I've simplified.

File: recipes/eleuther_eval.py [hunk 2]
```diff
@@ -112,15 +111,13 @@ def tok_batch_encode(
         tokenized_text = [self.tok_encode(x) for x in text]
 
         # pad left
-        x = pad_sequence(
+        x = left_pad_sequence(
             [
                 torch.tensor(x[::-1]) for x in tokenized_text
```
Comment (SalmanMohammadi) [inline recipes/eleuther_eval.py:116]: whoops gotta fix

File: torchtune/data/_collate.py [hunk 3]
```diff
@@ -12,7 +12,136 @@
...
+                              [-10,   4,   5,   6,   7],
+                              [  8,   9,  10,  11,  12]])
+        }
+    """
+    if not isinstance(keys_to_pad, list) or not keys_to_pad:
+        raise ValueError(
+            f"keys_to_pad should be a list of strings with at least one element, but found {keys_to_pad}!"
+        )
+    if isinstance(padding_idx, dict):
+        if isinstance(padding_idx, dict) and not set(padding_idx.keys()) == set(
```
Comment (RdoubleA) [inline torchtune/data/_collate.py:118]: do you need to check for dict twice?

File: torchtune/data/_collate.py [hunk 4]
```diff
@@ -12,7 +12,136 @@
...
+            )
+        if not set(keys_to_pad) <= set(batch[0].keys()):
+            raise ValueError(
+                "keys_to_pad should be a subset of keys in the batch, but found "
+                f"{set(keys_to_pad)} and {set(batch[0].keys())}, respectively."
+            )
+
+    batch_keys = [k for k in batch[0].keys() if k not in keys_to_pad]
+    output_dict = {k: torch.tensor([x[k] for x in batch]) for k in batch_keys}
+    for k in keys_to_pad:
```
Comment (RdoubleA) [inline torchtune/data/_collate.py:133]: maybe should convert to a set to avoid accidental duplicates

File: torchtune/data/_collate.py [hunk 5]
```diff
@@ -12,7 +12,136 @@
...
+                f"padding_idx was provided as a dictionary, but the keys ({padding_idx.keys()}) "
+                f"are not the same as keys_to_pad ({keys_to_pad})"
+            )
+        if not set(keys_to_pad) <= set(batch[0].keys()):
+            raise ValueError(
+                "keys_to_pad should be a subset of keys in the batch, but found "
+                f"{set(keys_to_pad)} and {set(batch[0].keys())}, respectively."
+            )
+
+    batch_keys = [k for k in batch[0].keys() if k not in keys_to_pad]
```
Comment (RdoubleA) [inline torchtune/data/_collate.py:131]: a quick comment explaining that this is taking out keys that won't be padded would be good

File: torchtune/data/_collate.py [hunk 6]
```diff
@@ -12,7 +12,136 @@
...
+        map(lambda x: torch.flip(x, dims=[0]), sequences),
+        batch_first=batch_first,
+        padding_value=padding_value,
+    ).flip(dims=[1])
+
+
 def padded_collate(
+    batch: List[Dict[str, List[int]]],
+    *,
+    pad_fn: Callable,
```
Comment (RdoubleA) [inline torchtune/data/_collate.py:59]: I'm still on the fence about this, it feels like a case of generalizing too early. In PPO you are only using left pad. I imagine right pad would also be a common case, but why not just specify with a string "left" or "right"?

**Review by RdoubleA — APPROVED** [review]
beautiful
